nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java

```diff
@@ -97,25 +98,48 @@ public static void checkInstantiationForParameterizedTypedTree(
     if (baseType == null) {
       return;
     }
+    boolean[] typeParamsWithNullableUpperBound =
+        getTypeParamsWithNullableUpperBound(baseType, config, state, handler);
     com.sun.tools.javac.util.List<Type> baseTypeArgs = baseType.tsym.type.getTypeArguments();
     for (int i = 0; i < baseTypeArgs.size(); i++) {
-      if (nullableTypeArguments.containsKey(i)) {
-
-        Type typeVariable = baseTypeArgs.get(i);
-        Type upperBound = typeVariable.getUpperBound();
-        com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrors =
-            upperBound.getAnnotationMirrors();
-        boolean hasNullableAnnotation =
-            Nullness.hasNullableAnnotation(annotationMirrors.stream(), config)
-                || handler.onOverrideTypeParameterUpperBound(baseType.tsym.toString(), i);
-        // if base type argument does not have @Nullable annotation then the instantiation is
+      if (nullableTypeArguments.containsKey(i) && !typeParamsWithNullableUpperBound[i]) {
+        // if base type variable does not have @Nullable upper bound then the instantiation is
         // invalid
-        if (!hasNullableAnnotation) {
-          reportInvalidInstantiationError(
-              nullableTypeArguments.get(i), baseType, typeVariable, state, analysis);
+        reportInvalidInstantiationError(
+            nullableTypeArguments.get(i), baseType, baseTypeArgs.get(i), state, analysis);
+      }
+    }
+  }
+
+  private static boolean[] getTypeParamsWithNullableUpperBound(
+      Type type, Config config, VisitorState state, Handler handler) {
+    Symbol.TypeSymbol tsym = type.tsym;
+    com.sun.tools.javac.util.List<Type> baseTypeArgs = tsym.type.getTypeArguments();
+    boolean[] result = new boolean[baseTypeArgs.size()];
+    for (int i = 0; i < baseTypeArgs.size(); i++) {
+      Type typeVariable = baseTypeArgs.get(i);
+      Type upperBound = typeVariable.getUpperBound();
+      com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrors =
+          upperBound.getAnnotationMirrors();
+      if (Nullness.hasNullableAnnotation(annotationMirrors.stream(), config)
+          || handler.onOverrideTypeParameterUpperBound(type.tsym.toString(), i)) {
+        result[i] = true;
+      }
+    }
+    // for handling bytecodes
```

Comment:
The code below is the new part. Maybe slightly more context here on why it's needed? (I know the difference between getting type info for source vs bytecode, but it won't be obvious to someone looking at this code why this is needed or where to find more info :) )

Reply:
Thanks, fixed in 8b61b6aad1c028e3078aaaf054bc05b164893edb